### PR TITLE
[feat]: [CCM-9628]: API support for prefilling 12 monthly budget amounts

### PR DESCRIPTION
**Pull request by iamanmolm**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/39201)

### CODE REVIEW

**Comment by iamanmolm** [comment]
trigger messagecheck

File: ce-nextgen/service/src/main/java/io/harness/ccm/remote/resources/perspectives/PerspectiveResource.java [hunk 1]
```diff
@@ -195,6 +197,45 @@ public PerspectiveResource(CEViewService ceViewService, CEReportScheduleService
...
+      BudgetPeriod period,
+      @NotNull @Valid @QueryParam("type") @Parameter(required = true,
+          description =
+              "The type (SPECIFIED_AMOUNT, PREVIOUS_MONTH_SPEND, PREVIOUS_PERIOD_SPEND, MONTHLY) for which we want the cost")
+      BudgetType type,
+      @NotNull @Valid @QueryParam("breakdown") @Parameter(required = true,
+          description = "The breakdown (MONTHLY, YEARLY) for which we want the cost") BudgetBreakdown breakdown) {
+    rbacHelper.checkPerspectiveViewPermission(accountId, null, null);
+    Double[] response = null;
+    if (period == BudgetPeriod.YEARLY && type == BudgetType.PREVIOUS_MONTH_SPEND
```
Comment (iamanmolm) [inline ce-nextgen/service/src/main/java/io/harness/ccm/remote/resources/perspectives/PerspectiveResource.java:232]: BudgetType.Previous_Period_Spend

File: 374-ceng-graphql/src/main/java/io/harness/ccm/graphql/core/budget/BudgetCostServiceImpl.java [hunk 2]
```diff
@@ -123,6 +124,28 @@ public double getLastPeriodCost(String accountId, String perspectiveId, long sta
...
+      String accountId, String perspectiveId, long startTime, BudgetPeriod period, BudgetType type) {
+    Double[] result;
+    String cloudProviderTableName = bigQueryHelper.getCloudProviderTableName(accountId, UNIFIED_TABLE);
+    List<QLCEViewGroupBy> groupBy = new ArrayList<>();
+    groupBy.add(
+        QLCEViewGroupBy.builder()
+            .timeTruncGroupBy(
+                QLCEViewTimeTruncGroupBy.builder().resolution(getTimeResolutionForBudget(BudgetPeriod.MONTHLY)).build())
+            .build());
+    long endTime = startTime - BudgetUtils.ONE_DAY_MILLIS;
```
Comment (iamanmolm) [inline 374-ceng-graphql/src/main/java/io/harness/ccm/graphql/core/budget/BudgetCostServiceImpl.java:137]: Accordingly change the endTime as well.

**Comment by iamanmolm** [comment]
trigger compile, ti0, ti1, ti2, ti3, ti4

**Comment by iamanmolm** [comment]
trigger compile, ti0, ti1, ti2, ti3, ti4

**Comment by iamanmolm** [comment]
trigger ti1, ti2

**Comment by iamanmolm** [comment]
trigger compile, featurenamecheck, ft1, ft2

**Review by svmav — DISMISSED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39201)* status: 9 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[`374-ceng-graphql/src/main/java/io/harness/ccm/graphql/core/budget/BudgetCostServiceImpl.java` line 127 at r3](https://reviewable.io/reviews/harness/harness-core/39201#-NFvhono8_tPyCJwBv6q:-NFvhono8_tPyCJwBv6r:b-3m3rrj) ([raw file](https://github.com/harness/harness-core/blob/6121d6a2ab4354d8d5273d49a4510fd2d4fdc7fe/374-ceng-graphql/src/main/java/io/harness/ccm/graphql/core/budget/BudgetCostServiceImpl.java#L127)):*
> ```Java
>   }
> 
>   public Double[] getLastYearMonthlyCost(
> ```

Is BudgetType used anywhere?
___
*[`450-ce-views/src/main/java/io/harness/ccm/views/service/impl/ViewsBillingServiceImpl.java` line 720 at r3](https://reviewable.io/reviews/harness/harness-core/39201#-NFviE6e6nS4Me2fHKKR:-NFviE6e6nS4Me2fHKKS:b-gihgmq) ([raw file](https://github.com/harness/harness-core/blob/6121d6a2ab4354d8d5273d49a4510fd2d4fdc7fe/450-ce-views/src/main/java/io/harness/ccm/views/service/impl/ViewsBillingServiceImpl.java#L720)):*
> ```Java
> 
>     Double[] monthlyCosts = new Double[months];
>     int i = ((Long) result.getTotalRows()).intValue();
> ```

Can we have some proper variable name instead of i

**Review by iamanmolm — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39201)* status: 9 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`374-ceng-graphql/src/main/java/io/harness/ccm/graphql/core/budget/BudgetCostServiceImpl.java` line 127 at r3](https://reviewable.io/reviews/harness/harness-core/39201#-NFvhono8_tPyCJwBv6q:-NFvqMr9DRBv5SDbOorD:b-lrjkk4) ([raw file](https://github.com/harness/harness-core/blob/6121d6a2ab4354d8d5273d49a4510fd2d4fdc7fe/374-ceng-graphql/src/main/java/io/harness/ccm/graphql/core/budget/BudgetCostServiceImpl.java#L127)):*
<details><summary><i>Previously, svmav (Shubhanshu Verma) wrote…</i></summary><blockquote>

Is BudgetType used anywhere?
</blockquote></details>

Done as part of forecasting PR
___
*[`450-ce-views/src/main/java/io/harness/ccm/views/service/impl/ViewsBillingServiceImpl.java` line 720 at r3](https://reviewable.io/reviews/harness/harness-core/39201#-NFviE6e6nS4Me2fHKKR:-NFvqYl89pElTJbKkwMl:b-wd2rqj) ([raw file](https://github.com/harness/harness-core/blob/6121d6a2ab4354d8d5273d49a4510fd2d4fdc7fe/450-ce-views/src/main/java/io/harness/ccm/views/service/impl/ViewsBillingServiceImpl.java#L720)):*
<details><summary><i>Previously, svmav (Shubhanshu Verma) wrote…</i></summary><blockquote>

Can we have some proper variable name instead of i
</blockquote></details>

Will fix it in forecasting PR

**Review by svmav — COMMENTED** [review]
> Reviewed 4 of 7 files at r1, 2 of 5 files at r2, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39201)* status: 2. unresolved discussions. 0 of 1 approvals obtained.

**Review by svmav — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39201)* status: :shipit: complete! ✓

**Pull request merged**